***
### Import required Libraries

In [1]:
# Data processing
import numpy as np
import pandas as pd

# Visualization 
import seaborn as sns
import matplotlib.pyplot as plt

# to preprocess the data 
from mlxtend.preprocessing import TransactionEncoder

# for Market Basket Analysis 
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

%matplotlib inline

In [2]:
# Load churn dataset into Pandas dataframe 

path = r"C:\Users\Abhishek\Desktop\D212\Task 3\teleco_market_basket.csv"
 
df_telco = pd.read_csv(path)  
df_telco.head()   

Item01                            Item02  \
0                                    NaN                               NaN   
1           Logitech M510 Wireless mouse                         HP 63 Ink   
2                                    NaN                               NaN   
3  Apple Lightning to Digital AV Adapter  TP-Link AC1750 Smart WiFi Router   
4                                    NaN                               NaN   

         Item03                      Item04                     Item05  \
0           NaN                         NaN                        NaN   
1     HP 65 ink  nonda USB C to USB Adapter  10ft iPHone Charger Cable   
2           NaN                         NaN                        NaN   
3  Apple Pencil                         NaN                        NaN   
4           NaN                         NaN                        NaN   

         Item06                        Item07  \
0           NaN                           NaN   
1  HP 902XL ink  Creative Pebble 2.0 Speakers   
2           NaN                           NaN   
3           NaN                           NaN   
4           NaN                           NaN   

                                Item08                         Item09  \
0                                  NaN                            NaN   
1  Cleaning Gel Universal Dust Cleaner  Micro Center 32GB Memory card   
2                                  NaN                            NaN   
3                                  NaN                            NaN   
4                                  NaN                            NaN   

                                    Item10                        Item11  \
0                                      NaN                           NaN   
1  YUNSONG 3pack 6ft Nylon Lightning Cable  TopMate C5 Laptop Cooler pad   
2                                      NaN                           NaN   
3                                      NaN                           NaN   
4                                      NaN                           NaN   

                      Item12                        Item13  \
0                        NaN                           NaN   
1  Apple USB-C Charger cable  HyperX Cloud Stinger Headset   
2                        NaN                           NaN   
3                        NaN                           NaN   
4                        NaN                           NaN   

                        Item14                          Item15  \
0                          NaN                             NaN   
1  TONOR USB Gaming Microphone  Dust-Off Compressed Gas 2 pack   
2                          NaN                             NaN   
3                          NaN                             NaN   
4                          NaN                             NaN   

                           Item16                 Item17  \
0                             NaN                    NaN   
1  3A USB Type C Cable 3 pack 6FT  HOVAMP iPhone charger   
2                             NaN                    NaN   
3                             NaN                    NaN   
4                             NaN                    NaN   

                     Item18                                Item19  \
0                       NaN                                   NaN   
1  SanDisk Ultra 128GB card  FEEL2NICE 5 pack 10ft Lighning cable   
2                       NaN                                   NaN   
3                       NaN                                   NaN   
4                       NaN                                   NaN   

                                Item20  
0                                  NaN  
1  FEIYOLD Blue light Blocking Glasses  
2                                  NaN  
3                                  NaN  
4                                  NaN

In [3]:
df_telco.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15002 entries, 0 to 15001
Data columns (total 20 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Item01  7501 non-null   object
 1   Item02  5747 non-null   object
 2   Item03  4389 non-null   object
 3   Item04  3345 non-null   object
 4   Item05  2529 non-null   object
 5   Item06  1864 non-null   object
 6   Item07  1369 non-null   object
 7   Item08  981 non-null    object
 8   Item09  654 non-null    object
 9   Item10  395 non-null    object
 10  Item11  256 non-null    object
 11  Item12  154 non-null    object
 12  Item13  87 non-null     object
 13  Item14  47 non-null     object
 14  Item15  25 non-null     object
 15  Item16  8 non-null      object
 16  Item17  4 non-null      object
 17  Item18  4 non-null      object
 18  Item19  3 non-null      object
 19  Item20  1 non-null      object
dtypes: object(20)
memory usage: 2.3+ MB


In [4]:
df_telco.isnull()

Item01  Item02  Item03  Item04  Item05  Item06  Item07  Item08  Item09  \
0        True    True    True    True    True    True    True    True    True   
1       False   False   False   False   False   False   False   False   False   
2        True    True    True    True    True    True    True    True    True   
3       False   False   False    True    True    True    True    True    True   
4        True    True    True    True    True    True    True    True    True   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
14997   False    True    True    True    True    True    True    True    True   
14998    True    True    True    True    True    True    True    True    True   
14999   False   False    True    True    True    True    True    True    True   
15000    True    True    True    True    True    True    True    True    True   
15001   False   False   False   False    True    True    True    True    True   

       Item10  Item11  Item12  Item13  Item14  Item15  Item16  Item17  Item18  \
0        True    True    True    True    True    True    True    True    True   
1       False   False   False   False   False   False   False   False   False   
2        True    True    True    True    True    True    True    True    True   
3        True    True    True    True    True    True    True    True    True   
4        True    True    True    True    True    True    True    True    True   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
14997    True    True    True    True    True    True    True    True    True   
14998    True    True    True    True    True    True    True    True    True   
14999    True    True    True    True    True    True    True    True    True   
15000    True    True    True    True    True    True    True    True    True   
15001    True    True    True    True    True    True    True    True    True   

       Item19  Item20  
0        True    True  
1       False   False  
2        True    True  
3        True    True  
4        True    True  
...       ...     ...  
14997    True    True  
14998    True    True  
14999    True    True  
15000    True    True  
15001    True    True  

[15002 rows x 20 columns]

In [5]:
df_telco.isnull().sum()

Item01     7501
Item02     9255
Item03    10613
Item04    11657
Item05    12473
Item06    13138
Item07    13633
Item08    14021
Item09    14348
Item10    14607
Item11    14746
Item12    14848
Item13    14915
Item14    14955
Item15    14977
Item16    14994
Item17    14998
Item18    14998
Item19    14999
Item20    15001
dtype: int64

In [6]:
# set the max columns
pd.set_option('display.max_columns',None)   
df_telco.head()

Item01                            Item02  \
0                                    NaN                               NaN   
1           Logitech M510 Wireless mouse                         HP 63 Ink   
2                                    NaN                               NaN   
3  Apple Lightning to Digital AV Adapter  TP-Link AC1750 Smart WiFi Router   
4                                    NaN                               NaN   

         Item03                      Item04                     Item05  \
0           NaN                         NaN                        NaN   
1     HP 65 ink  nonda USB C to USB Adapter  10ft iPHone Charger Cable   
2           NaN                         NaN                        NaN   
3  Apple Pencil                         NaN                        NaN   
4           NaN                         NaN                        NaN   

         Item06                        Item07  \
0           NaN                           NaN   
1  HP 902XL ink  Creative Pebble 2.0 Speakers   
2           NaN                           NaN   
3           NaN                           NaN   
4           NaN                           NaN   

                                Item08                         Item09  \
0                                  NaN                            NaN   
1  Cleaning Gel Universal Dust Cleaner  Micro Center 32GB Memory card   
2                                  NaN                            NaN   
3                                  NaN                            NaN   
4                                  NaN                            NaN   

                                    Item10                        Item11  \
0                                      NaN                           NaN   
1  YUNSONG 3pack 6ft Nylon Lightning Cable  TopMate C5 Laptop Cooler pad   
2                                      NaN                           NaN   
3                                      NaN                           NaN   
4                                      NaN                           NaN   

                      Item12                        Item13  \
0                        NaN                           NaN   
1  Apple USB-C Charger cable  HyperX Cloud Stinger Headset   
2                        NaN                           NaN   
3                        NaN                           NaN   
4                        NaN                           NaN   

                        Item14                          Item15  \
0                          NaN                             NaN   
1  TONOR USB Gaming Microphone  Dust-Off Compressed Gas 2 pack   
2                          NaN                             NaN   
3                          NaN                             NaN   
4                          NaN                             NaN   

                           Item16                 Item17  \
0                             NaN                    NaN   
1  3A USB Type C Cable 3 pack 6FT  HOVAMP iPhone charger   
2                             NaN                    NaN   
3                             NaN                    NaN   
4                             NaN                    NaN   

                     Item18                                Item19  \
0                       NaN                                   NaN   
1  SanDisk Ultra 128GB card  FEEL2NICE 5 pack 10ft Lighning cable   
2                       NaN                                   NaN   
3                       NaN                                   NaN   
4                       NaN                                   NaN   

                                Item20  
0                                  NaN  
1  FEIYOLD Blue light Blocking Glasses  
2                                  NaN  
3                                  NaN  
4                                  NaN

### Prepare the Dataset

In the code below the dropna() is used to remove rows with NaN values. The axis=0 parameter specifies that rows should be dropped, and how='all' indicates that a row will only be dropped if all its values are NaN.

In [7]:
# remove the rows which has all null values 
df_telco = df_telco.dropna(axis=0, how='all')

In [8]:
df_telco.head()

Item01  \
1               Logitech M510 Wireless mouse   
3      Apple Lightning to Digital AV Adapter   
5  UNEN Mfi Certified 5-pack Lightning Cable   
7                        Cat8 Ethernet Cable   
9             Dust-Off Compressed Gas 2 pack   

                             Item02                      Item03  \
1                         HP 63 Ink                   HP 65 ink   
3  TP-Link AC1750 Smart WiFi Router                Apple Pencil   
5                               NaN                         NaN   
7                         HP 65 ink                         NaN   
9     Screen Mom Screen Cleaner kit  Moread HDMI to VGA Adapter   

                       Item04                     Item05        Item06  \
1  nonda USB C to USB Adapter  10ft iPHone Charger Cable  HP 902XL ink   
3                         NaN                        NaN           NaN   
5                         NaN                        NaN           NaN   
7                         NaN                        NaN           NaN   
9       HP 62XL Tri-Color ink  Apple USB-C Charger cable           NaN   

                         Item07                               Item08  \
1  Creative Pebble 2.0 Speakers  Cleaning Gel Universal Dust Cleaner   
3                           NaN                                  NaN   
5                           NaN                                  NaN   
7                           NaN                                  NaN   
9                           NaN                                  NaN   

                          Item09                                   Item10  \
1  Micro Center 32GB Memory card  YUNSONG 3pack 6ft Nylon Lightning Cable   
3                            NaN                                      NaN   
5                            NaN                                      NaN   
7                            NaN                                      NaN   
9                            NaN                                      NaN   

                         Item11                     Item12  \
1  TopMate C5 Laptop Cooler pad  Apple USB-C Charger cable   
3                           NaN                        NaN   
5                           NaN                        NaN   
7                           NaN                        NaN   
9                           NaN                        NaN   

                         Item13                       Item14  \
1  HyperX Cloud Stinger Headset  TONOR USB Gaming Microphone   
3                           NaN                          NaN   
5                           NaN                          NaN   
7                           NaN                          NaN   
9                           NaN                          NaN   

                           Item15                          Item16  \
1  Dust-Off Compressed Gas 2 pack  3A USB Type C Cable 3 pack 6FT   
3                             NaN                             NaN   
5                             NaN                             NaN   
7                             NaN                             NaN   
9                             NaN                             NaN   

                  Item17                    Item18  \
1  HOVAMP iPhone charger  SanDisk Ultra 128GB card   
3                    NaN                       NaN   
5                    NaN                       NaN   
7                    NaN                       NaN   
9                    NaN                       NaN   

                                 Item19                               Item20  
1  FEEL2NICE 5 pack 10ft Lighning cable  FEIYOLD Blue light Blocking Glasses  
3                                   NaN                                  NaN  
5                                   NaN                                  NaN  
7                                   NaN                                  NaN  
9                                   NaN                                  NaN

In [9]:
df_telco.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7501 entries, 1 to 15001
Data columns (total 20 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Item01  7501 non-null   object
 1   Item02  5747 non-null   object
 2   Item03  4389 non-null   object
 3   Item04  3345 non-null   object
 4   Item05  2529 non-null   object
 5   Item06  1864 non-null   object
 6   Item07  1369 non-null   object
 7   Item08  981 non-null    object
 8   Item09  654 non-null    object
 9   Item10  395 non-null    object
 10  Item11  256 non-null    object
 11  Item12  154 non-null    object
 12  Item13  87 non-null     object
 13  Item14  47 non-null     object
 14  Item15  25 non-null     object
 15  Item16  8 non-null      object
 16  Item17  4 non-null      object
 17  Item18  4 non-null      object
 18  Item19  3 non-null      object
 19  Item20  1 non-null      object
dtypes: object(20)
memory usage: 1.2+ MB


In [10]:
df_telco.shape

(7501, 20)

### Convert Data to Transaction Format

Convert the dataset into a transaction format that is suitable for Market Basket Analysis. This involves creating a list of lists, where each inner list represents a transaction and contains the items present in that transaction.



In [11]:
# Create a list of list for each row in DataFrame
transactions = []
for i, transaction in df_telco.iterrows():
    transactions.append(transaction[transaction.notna()].tolist())

In [12]:
print ("Transaction 1:" , transactions[0])
print ('************************************************************************')
print ("Transaction 5:" , transactions[4])
print ('************************************************************************')
print ("Transaction 30:" , transactions[29])

Transaction 1: ['Logitech M510 Wireless mouse', 'HP 63 Ink', 'HP 65 ink', 'nonda USB C to USB Adapter', '10ft iPHone Charger Cable', 'HP 902XL ink', 'Creative Pebble 2.0 Speakers', 'Cleaning Gel Universal Dust Cleaner', 'Micro Center 32GB Memory card', 'YUNSONG 3pack 6ft Nylon Lightning Cable', 'TopMate C5 Laptop Cooler pad', 'Apple USB-C Charger cable', 'HyperX Cloud Stinger Headset', 'TONOR USB Gaming Microphone', 'Dust-Off Compressed Gas 2 pack', '3A USB Type C Cable 3 pack 6FT', 'HOVAMP iPhone charger', 'SanDisk Ultra 128GB card', 'FEEL2NICE 5 pack 10ft Lighning cable', 'FEIYOLD Blue light Blocking Glasses']
************************************************************************
Transaction 5: ['Dust-Off Compressed Gas 2 pack', 'Screen Mom Screen Cleaner kit', 'Moread HDMI to VGA Adapter', 'HP 62XL Tri-Color ink', 'Apple USB-C Charger cable']
************************************************************************
Transaction 30: ['Nylon Braided Lightning to USB cable', 'VIVO Dual

### Encode the Transaction Data

Encode the transaction data using the TransactionEncoder class from mlxtend.preprocessing. This will transform the transaction data into a one-hot encoded format.

In [13]:
# Instantiate transaction encoder 
te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)

# convert into DataFrame 
df_encoded = pd.DataFrame(te_ary, columns=te.columns_)

In [14]:
df_encoded

10ft iPHone Charger Cable  10ft iPHone Charger Cable 2 Pack  \
0                          True                             False   
1                         False                             False   
2                         False                             False   
3                         False                             False   
4                         False                             False   
...                         ...                               ...   
7496                      False                             False   
7497                      False                             False   
7498                      False                             False   
7499                      False                             False   
7500                      False                             False   

      3 pack Nylon Braided Lightning Cable  3A USB Type C Cable 3 pack 6FT  \
0                                    False                            True   
1                                    False                           False   
2                                    False                           False   
3                                    False                           False   
4                                    False                           False   
...                                    ...                             ...   
7496                                 False                           False   
7497                                 False                           False   
7498                                 False                           False   
7499                                 False                           False   
7500                                 False                           False   

      5pack Nylon Braided USB C cables  ARRIS SURFboard SB8200 Cable Modem  \
0                                False                               False   
1                                False                               False   
2                                False                               False   
3                                False                               False   
4                                False                               False   
...                                ...                                 ...   
7496                             False                               False   
7497                             False                                True   
7498                             False                               False   
7499                             False                               False   
7500                             False                               False   

      Anker 2-in-1 USB Card Reader  Anker 4-port USB hub  \
0                            False                 False   
1                            False                 False   
2                            False                 False   
3                            False                 False   
4                            False                 False   
...                            ...                   ...   
7496                         False                 False   
7497                         False                 False   
7498                         False                 False   
7499                         False                 False   
7500                         False                 False   

      Anker USB C to HDMI Adapter  Apple Lightning to Digital AV Adapter  \
0                           False                                  False   
1                           False                                   True   
2                           False                                  False   
3                           False                                  False   
4                           False                                  False   
...                           ...                                    ...   
7496                        False                 

### List items as columns to find if any empty columns

In [15]:
for col in df_encoded.columns:
    print(col)

10ft iPHone Charger Cable
10ft iPHone Charger Cable 2 Pack
3 pack Nylon Braided Lightning Cable
3A USB Type C Cable 3 pack 6FT
5pack Nylon Braided USB C cables
ARRIS SURFboard SB8200 Cable Modem
Anker 2-in-1 USB Card Reader
Anker 4-port USB hub
Anker USB C to HDMI Adapter
Apple Lightning to Digital AV Adapter
Apple Lightning to USB cable
Apple Magic Mouse 2
Apple Pencil
Apple Pencil 2nd Gen
Apple Power Adapter Extension Cable
Apple USB-C Charger cable
AutoFocus 1080p Webcam
BENGOO G90000 headset
Blue Light Blocking Glasses
Blue Light Blocking Glasses 2pack
Brother Genuine High Yield Toner Cartridge
Cat 6 Ethernet Cable 50ft
Cat8 Ethernet Cable
CicTsing MM057 2.4G Wireless Mouse
Cleaning Gel Universal Dust Cleaner
Creative Pebble 2.0 Speakers
DisplayPort ot HDMI adapter
Dust-Off Compressed Gas
Dust-Off Compressed Gas 2 pack
FEEL2NICE 5 pack 10ft Lighning cable
FEIYOLD Blue light Blocking Glasses
Falcon Dust Off Compressed Gas
HOVAMP Mfi 6pack Lightning Cable
HOVAMP iPhone charger
HP 61 

In [16]:
df_encoded.shape

(7501, 119)

### Prepared Dataset   

In [17]:
# Extract the clean data in CSV format 
df_encoded.to_csv('mba_prepared_data.csv', index = False, encoding = 'utf-8')

### Market Basket Analysis 

### Generate Frequent Itemsets

Use the apriori function from mlxtend.frequent_patterns to generate frequent itemsets from the encoded data.

In [18]:
frequent_itemsets = apriori(df_encoded, min_support=0.05, use_colnames=True)

In [19]:
frequent_itemsets

support                                           itemsets
0   0.050527                 (10ft iPHone Charger Cable 2 Pack)
1   0.068391                      (Anker USB C to HDMI Adapter)
2   0.087188            (Apple Lightning to Digital AV Adapter)
3   0.179709                                     (Apple Pencil)
4   0.132116                        (Apple USB-C Charger cable)
5   0.062525                              (Cat8 Ethernet Cable)
6   0.238368                   (Dust-Off Compressed Gas 2 pack)
7   0.065858              (FEIYOLD Blue light Blocking Glasses)
8   0.059992                   (Falcon Dust Off Compressed Gas)
9   0.163845                                        (HP 61 ink)
10  0.058526                            (HP 62XL Tri-Color ink)
11  0.079323                                      (HP 63XL Ink)
12  0.071457                     (Logitech M510 Wireless mouse)
13  0.095321             (Nylon Braided Lightning to USB cable)
14  0.051060                          (Premium Nylon USB Cable)
15  0.052393                            (SAMSUNG EVO 32GB card)
16  0.063325                         (SanDisk Ultra 128GB card)
17  0.098254                          (SanDisk Ultra 64GB card)
18  0.129583                    (Screen Mom Screen Cleaner kit)
19  0.095054                              (Stylus Pen for iPad)
20  0.081056                     (Syntech USB C to USB Adapter)
21  0.076523                     (TopMate C5 Laptop Cooler pad)
22  0.170911                            (USB 2.0 Printer cable)
23  0.080389                (USB Type C to USB-A Charger cable)
24  0.174110                 (VIVO Dual LCD Monitor Desk mount)
25  0.050927     (Apple Pencil, Dust-Off Compressed Gas 2 pack)
26  0.052660        (HP 61 ink, Dust-Off Compressed Gas 2 pack)
27  0.059725  (Dust-Off Compressed Gas 2 pack, VIVO Dual LCD...

In [20]:
# print number of items
print("There are",frequent_itemsets.shape[0], "sets of items." )

There are 28 sets of items.


### Generate Association Rules

Used the association_rules function from mlxtend.frequent_patterns to generate association rules from the frequent itemsets.

In [21]:
# Generate association rules
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.2)

In [22]:
rules

antecedents                         consequents  \
0                      (Apple Pencil)    (Dust-Off Compressed Gas 2 pack)   
1    (Dust-Off Compressed Gas 2 pack)                      (Apple Pencil)   
2                         (HP 61 ink)    (Dust-Off Compressed Gas 2 pack)   
3    (Dust-Off Compressed Gas 2 pack)                         (HP 61 ink)   
4    (Dust-Off Compressed Gas 2 pack)  (VIVO Dual LCD Monitor Desk mount)   
5  (VIVO Dual LCD Monitor Desk mount)    (Dust-Off Compressed Gas 2 pack)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.179709            0.238368  0.050927    0.283383  1.188845   
1            0.238368            0.179709  0.050927    0.213647  1.188845   
2            0.163845            0.238368  0.052660    0.321400  1.348332   
3            0.238368            0.163845  0.052660    0.220917  1.348332   
4            0.238368            0.174110  0.059725    0.250559  1.439085   
5            0.174110            0.238368  0.059725    0.343032  1.439085   

   leverage  conviction  zhangs_metric  
0  0.008090    1.062815       0.193648  
1  0.008090    1.043158       0.208562  
2  0.013604    1.122357       0.308965  
3  0.013604    1.073256       0.339197  
4  0.018223    1.102008       0.400606  
5  0.018223    1.159314       0.369437

In [23]:
# Iterate over each rule and display the relevant information
for idx, rule in rules.iterrows():
    antecedents = ' + '.join(list(rule['antecedents']))
    consequents = ' + '.join(list(rule['consequents']))
    confidence = rule['confidence']
    support = rule['support']
    lift = rule['lift']
    
    print(f"Rule #{idx+1}: {antecedents} -> {consequents}")
    print(f"Confidence: {confidence:.2f}")
    print(f"Support: {support:.2f}")
    print(f"Lift: {lift:.2f}")
    print()

Rule #1: Apple Pencil -> Dust-Off Compressed Gas 2 pack
Confidence: 0.28
Support: 0.05
Lift: 1.19

Rule #2: Dust-Off Compressed Gas 2 pack -> Apple Pencil
Confidence: 0.21
Support: 0.05
Lift: 1.19

Rule #3: HP 61 ink -> Dust-Off Compressed Gas 2 pack
Confidence: 0.32
Support: 0.05
Lift: 1.35

Rule #4: Dust-Off Compressed Gas 2 pack -> HP 61 ink
Confidence: 0.22
Support: 0.05
Lift: 1.35

Rule #5: Dust-Off Compressed Gas 2 pack -> VIVO Dual LCD Monitor Desk mount
Confidence: 0.25
Support: 0.06
Lift: 1.44

Rule #6: VIVO Dual LCD Monitor Desk mount -> Dust-Off Compressed Gas 2 pack
Confidence: 0.34
Support: 0.06
Lift: 1.44



###  Top Three Rules 

The support matrix indicates the proportion of transactions in the dataset that contains a specific itemset. 

Higher support values indicate that the itemsets are popular and frequently co-purchased by customers. 

In [24]:
# Sort the rules based on support in descending order
rules_sorted_by_support = rules.sort_values(by='support', ascending=False)

In [25]:
# Select the top three rules with the highest lift
top_three_rules_support = rules_sorted_by_support.head(3)

In [26]:
top_three_rules_support 

antecedents                         consequents  \
4    (Dust-Off Compressed Gas 2 pack)  (VIVO Dual LCD Monitor Desk mount)   
5  (VIVO Dual LCD Monitor Desk mount)    (Dust-Off Compressed Gas 2 pack)   
2                         (HP 61 ink)    (Dust-Off Compressed Gas 2 pack)   

   antecedent support  consequent support   support  confidence      lift  \
4            0.238368            0.174110  0.059725    0.250559  1.439085   
5            0.174110            0.238368  0.059725    0.343032  1.439085   
2            0.163845            0.238368  0.052660    0.321400  1.348332   

   leverage  conviction  zhangs_metric  
4  0.018223    1.102008       0.400606  
5  0.018223    1.159314       0.369437  
2  0.013604    1.122357       0.308965

In [27]:
# Iterate over each rule in the DataFrame
for index, rule in top_three_rules_support.iterrows():
    antecedents = ', '.join(list(rule['antecedents']))
    consequents = ', '.join(list(rule['consequents']))
    confidence = rule['confidence']
    support = rule['support']
    lift = rule['lift']

    # Print the rule and its corresponding metrics
    print(f"Rule: {antecedents} -> {consequents}")
    print(f"Confidence: {confidence:.2f}")
    print(f"Support: {support:.2f}")
    print(f"Lift: {lift:.2f}")
    print()

Rule: Dust-Off Compressed Gas 2 pack -> VIVO Dual LCD Monitor Desk mount
Confidence: 0.25
Support: 0.06
Lift: 1.44

Rule: VIVO Dual LCD Monitor Desk mount -> Dust-Off Compressed Gas 2 pack
Confidence: 0.34
Support: 0.06
Lift: 1.44

Rule: HP 61 ink -> Dust-Off Compressed Gas 2 pack
Confidence: 0.32
Support: 0.05
Lift: 1.35



**Rule 1** : 

Customers who purchased the "Dust-Off Compressed Gas 2 pack" have a 25% probability of also purchasing the "VIVO Dual LCD Monitor Desk mount". 


**Rule 2** : 

Customers who purchased the "VIVO Dual LCD Monitor Desk mount" have a 34% probability of also purchasing the "Dust-Off Compressed Gas 2 pack"


**Rule 3** : 

Customers who purchased the "HP 61 ink" have a 32% probability of also purchasing the "Dust-Off Compressed Gas 2 pack". 

References - 

Raschka, S. (2014). Installing mlxtend. 
http://rasbt.github.io/mlxtend/installation/
    
WGU, Resources. (n.d.). Market Basket Analysis webinar recording. 
https://wgu.hosted.panopto.com/Panopto/Pages/Viewer.aspx?id=dbe89ddb-e92f-4d40-a87a-af030178abf1